### DISCUSSION TITLE
Forward an event downstream when Graphcache applies optimistic updates

### DISCUSSION THREAD

**Discussion by DustinJSilk** [urql-graphql/urql#2752] [Q&A]
Hello!

I'm in the process of writing Urql bindings for [Qwik](https://qwik.builder.io/) and I'm trying to get optimistic updates working.

Qwik doesn't support traditional subscriptions because they can't be serialized (trust me ive tried). To make Urql work with Qwik, I wrote an exchange that basically triggers a new cache-only query any time new data is returned from the server. It means subscriptions aren't using Wonka.

To make optimistic mutations work, I need to know when there is new data available in the cache. The problem is, graphcache only sends data downstream if there is a subscriber registered which isn't possible with Qwik, and there's no way of knowing if it is an optimistic update either.

I haven't got any good solutions for this yet and I'm looking for some guidance.
What can I do, or can we get a change added to Urql?

Here is my repository:  https://github.com/DustinJSilk/qwik-urql

Thanks!

**Comment by JoviDeCroock** [comment]
I am not too familiar with `qwik` in general so bear with me here while I pose some clarifying questions, is it a no-go to hydrate the data on the client and to re-establish these subscriptions? What I am mainly asking here is why we aren't sending the data over to the client and using the `ssrExchange` to re-hydrate this, this means you can easily set up the subscriptions again without incurring a network request. This in turn means that any mutation would be able to do optimistic/... with your client.

I feel like not establishing the subscriptions is kind off against the whole functional streaming pattern we got going in urql

**Reply by meenie** [reply]
Qwik does not have a typical re-hydration stage. When the app loads on the browser, no JS is loaded (except for a ~1kb inline script called `QwikLoader` that helps with "resuming" the app when an interaction occurs).  The interactions, i.e. click handlers, only load the JS required to perform the action required. The handlers are loaded via an async `import` and then provided its lexical scope of variables to do what it needs.  So the variables that you use within these handlers need to be primitives that can be serialized into JSON. There are some objects that have special cases for them like `Date` or `Promise`.  

With that said, the goal is to make the GraphQL queries on the server side, serialize everything into HTML, pause the application, and then load it into the browser. When an interaction happens that causes a mutation, that will "wake up" just that part of that app, load Urql and whatever other libraries it needs, to perform the mutation. The other places in the app where you have done `useQuery` should be told, "hey, you have some new data, you should now update from the Urql Cache" (and if done correctly, doesn't even need to load the JS for those components to update themselves (it's nuts)).  At the moment, this works, but only after the mutation response comes back and updates the cache.  What we are trying to achieve is optimistic updates.  You can see the exchange that @DustinJSilk created here: https://github.com/DustinJSilk/qwik-urql/blob/f694d9acfd3e6e149feb4a8b2ca09959a31f1a6b/src/exchange/qwik-exchange.ts

**Reply by kitten** [reply]
Ok, so as far as I can tell, the issue then starts with knowing which queries to update. Is that right?

The problem here is that `urql`'s `Client` has a concept of "if no one's listening who will care about the result?", so in that sense, there's a bit of a problem here with what comes first; the listening or the update?

The `this.cache.dependencies` you've got in your code sample is actually extremely similar to how we determine which queries are affected in Graphache — the difference obviously being, any query must be traversed fully and we've got the `keys` config, which all alter how the keys are used and tracked as dependencies (including updates we can't see by mutations due to the `updates` config, etc, etc)

In short, the non-optimistic mutation updates in `urql` (and the optimistic ones in Graphcache) work because we can tell the `Client` that a query should be "reexecuted" against the cache.
As you noted then, the problem is that this must be picked up somehow.

So, I'm basically curious:
- If I understand this correctly, you've got this set up in the reverse, right? First you execute the mutation, then you update the query results passively, right? How do the "inactive" queries know to accept the result?
- Would it be possible to do the reverse and "wake up" all queries against serialised cache results (i.e. how the `ssrExchange` functions) when a mutation is triggered first, to let the normalised cache establish dependencies?

**Reply by DustinJSilk** [reply]
This has all been very helpful and thanks for the quick replies everyone!

I've managed to make some progress waking up subscriptions in [this pr.](https://github.com/DustinJSilk/qwik-urql/pull/16). 

I havent managed to get optimistic updates working yet, i'm having trouble getting Qwik to render the optimistic updates, although I am receiving them in the subscription now. Im not entirely sure if its possible yet, i'll keep trying and report back.

**EDIT:** 

I'll try and answer your questions:

- I believe thats correct, originally, I'd re-execute queries by checking the response data. Urql and graphcache would then resolve those queries like normal requests.
- I've done this now in a new branch linked above. There's 2 issues preventing this from working: 
1. Its not possible to know which subscription to wake up before getting results back because the result could contain any document. The best I can do is wake up all subscriptions that exist for the expected __typename results. In which case, watching results shouldn't be the default behaviour but rather an opt in since it has a negative performance impact according to Qwiks design. This would require having access to the schema, or all queries would need to be woken up, which doesn't sound ideal either.
2. Qwik seems to be blocking UI updates returned from the subscription which is preventing optimistic updates from working. Still no work around.

**Reply by DustinJSilk** [reply]
I've managed to wake up subscriptions properly now and they're working with optimistic responses. The final piece of the puzzle is now figuring out which subscriptions to wake up before a request is sent.
